language/move-prover/bytecode-to-boogie/src/env.rs

```diff
@@ -0,0 +1,726 @@
+        let field_data = if let StructFieldInformation::Declared {
+            field_count,
+            fields,
+        } = module.struct_def_at(def_idx).field_information
+        {
+            (fields.0..fields.0 + field_count)
+                .map(|idx| FieldData {
+                    def_idx: FieldDefinitionIndex(idx),
+                })
+                .collect()
+        } else {
+            vec![]
+        };
+
+        StructData {
+            def_idx,
+            handle_idx,
+            field_data,
+        }
+    }
+
+    /// Adds a new module to the environment. StructData and FunctionData need to be provided
+    /// in definition index order.
+    pub fn add(
+        &mut self,
+        module: VerifiedModule,
+        source_map: ModuleSourceMap<Loc>,
+        struct_data: Vec<StructData>,
+        function_data: Vec<FunctionData>,
+    ) {
+        let idx = self.module_data.len();
+        let name = module
+            .identifier_at(module.module_handle_at(ModuleHandleIndex(0 as u16)).name)
+            .to_owned();
+        // TODO: verify whether the first element in the pool is actually the modules address.
+        let address = module.address_pool()[0];
+        self.module_data.push(ModuleData {
+            id: ModuleId::new(address, name),
+            idx,
+            module,
+            struct_data,
+            function_data,
+            source_map,
+        });
+    }
+
+    /// Finds a module by name and returns an environment for it.
+    pub fn find_module<'env>(&'env self, id: &ModuleId) -> Option<ModuleEnv<'env>> {
+        for module_data in &self.module_data {
+            let module_env = ModuleEnv {
+                env: self,
+                data: module_data,
+            };
+            if module_env.get_id() == id {
+                return Some(module_env);
+            }
+        }
+        None
+    }
+
+    /// Gets a module by index.
+    pub fn get_module<'env>(&'env self, idx: usize) -> ModuleEnv<'env> {
+        let module_data = &self.module_data[idx];
+        ModuleEnv {
+            env: self,
+            data: module_data,
+        }
+    }
+
+    /// Returns an iterator for all modules in the environment.
+    pub fn get_modules<'env>(&'env self) -> impl Iterator<Item = ModuleEnv<'env>> {
+        self.module_data.iter().map(move |module_data| ModuleEnv {
+            env: self,
+            data: module_data,
+        })
+    }
+
+    /// Returns an iterator for all bytecode modules in the environment.
+    pub fn get_bytecode_modules<'env>(&'env self) -> impl Iterator<Item = &'env VerifiedModule> {
+        self.module_data
+            .iter()
+            .map(|module_data| &module_data.module)
+    }
+}
+
+/// # Module Environment
+
+/// Represents data for a module.
+#[derive(Debug)]
+pub struct ModuleData {
+    /// Module id (pair of address and name)
+    id: ModuleId,
+
+    /// Index of this module in the global env.
+    idx: usize,
```

Comment:
A type alias for this might improve readability (but on the other hand might be overengineering)

Reply:
Actually had the alias already, but forget to use it here ;-)